### PR TITLE
split contracts reimbursement and capitation logic, refactored contracts module

### PR DESCRIPTION
**Pull request by PavelVesnin**

### CODE REVIEW

File: apps/core/lib/core/contract_requests/validator.ex [hunk 1]
```diff
@@ -545,7 +545,9 @@ defmodule Core.ContractRequests.Validator do
   end
 
   def validate_contract_request_content(:sign, %RequestPack{type: @reimbursement} = pack, _client_id) do
-    with medical_program <- MedicalPrograms.get_by_id(pack.decoded_content["medical_program_id"]),
+    medical_program_id = pack.decoded_content["medical_program_id"] || pack.contract_request.medical_program_id
+
+    with medical_program <- MedicalPrograms.get_by_id(medical_program_id),
```
Comment (progsmile) [inline apps/core/lib/core/contract_requests/validator.ex]: matches on everything

File: apps/core/lib/core/contracts/contracts.ex [hunk 2]
```diff
@@ -92,15 +55,24 @@ defmodule Core.Contracts do
...
-  def create(%{parent_contract_id: parent_contract_id} = params, user_id) when not is_nil(parent_contract_id) do
-    with %CapitationContract{} = contract <- PRMRepo.get(CapitationContract, parent_contract_id),
+  def create_from_contract_request(
+        %RequestPack{contract_request: %{parent_contract_id: parent_contract_id}} = pack,
+        user_id
+      )
+      when not is_nil(parent_contract_id) do
+    schema = get_contract_schema_relatively_contract_request(pack.schema)
+
+    with contract <- PRMRepo.get(schema, parent_contract_id),
```
Comment (progsmile) [inline apps/core/lib/core/contracts/contracts.ex]: same

File: apps/core/lib/core/contracts/contracts.ex [hunk 3]
```diff
@@ -777,7 +524,7 @@ defmodule Core.Contracts do
     |> PRMRepo.preload(:contract_divisions)
   end
 
-  defp get_bucket do
-    Confex.fetch_env!(:core, Core.API.MediaStorage)[:contract_bucket]
+  defp load_references(%ReimbursementContract{} = contract) do
```
Comment (progsmile) [inline apps/core/lib/core/contracts/contracts.ex]: there can be match error on capitation contract

File: apps/core/lib/core/contracts/contracts/capitation_contract.ex [hunk 4]
```diff
@@ -18,4 +53,48 @@ defmodule Core.Contracts.CapitationContract do
...
+    attrs =
+      case Map.get(attrs, :contractor_employee_divisions) do
+        nil ->
+          attrs
+
+        contractor_employee_divisions ->
+          contractor_employee_divisions =
+            Enum.map(
+              contractor_employee_divisions,
+              &(&1
```
Comment (progsmile) [inline apps/core/lib/core/contracts/contracts/capitation_contract.ex]: Map.merge

File: apps/ehealth/test/web/controllers/contract_request/reimbursement_controller_test.exs [hunk 5]
```diff
@@ -24,13 +25,30 @@ defmodule EHealth.Web.ContractRequest.ReimbursementControllerTest do
...
+      assert 4 == length(resp_data)
+    end
+
     test "successfully finds by medical_program_id", %{conn: conn} do
       nhs()
 
       %{id: medical_program_id} = insert(:prm, :medical_program)
       insert_list(2, :il, :reimbursement_contract_request, medical_program_id: medical_program_id)
       insert_list(4, :il, :reimbursement_contract_request)
-      insert_list(8, :il, :capitation_contract_request)
```
Comment (progsmile) [inline apps/ehealth/test/web/controllers/contract_request/reimbursement_controller_test.exs]: `:capitation_contract_request`s were inserted intentfully

File: apps/ehealth/test/web/controllers/contract_request/reimbursement_controller_test.exs [hunk 6]
```diff
@@ -377,9 +395,773 @@ defmodule EHealth.Web.ContractRequest.ReimbursementControllerTest do
...
+               |> put_client_id_header(client_id)
+               |> put_consumer_id_header(user_id)
+               |> put_req_header("msp_drfo", legal_entity.edrpou)
+               |> patch(contract_request_path(conn, :sign_msp, @path_type, contract_request.id), %{
+                 "signed_content" => data |> Poison.encode!() |> Base.encode64(),
+                 "signed_content_encoding" => "base64"
+               })
+               |> json_response(200)
+
+      schema =
```
Comment (progsmile) [inline apps/ehealth/test/web/controllers/contract_request/reimbursement_controller_test.exs]: use `assert_show_response_schema`

Comment (progsmile) [reply apps/ehealth/test/web/controllers/contract_request/reimbursement_controller_test.exs]: suppose it's legacy, with Poison.decode

File: apps/core/lib/core/contracts/contracts.ex (hunk 3 above)
Comment (PavelVesnin) [reply apps/core/lib/core/contracts/contracts.ex]: `defp load_references(%CapitationContract{} = contract) do` <- this function exists

Comment (progsmile) [reply apps/core/lib/core/contracts/contracts.ex]: щл

Comment (PavelVesnin) [reply apps/core/lib/core/contracts/contracts.ex]: you mean `ok`? :)

File: apps/core/lib/core/contract_requests/contract_requests.ex [hunk 7]
```diff
@@ -6,17 +6,17 @@ defmodule Core.ContractRequests do
   import Core.API.Helpers.Connection, only: [get_consumer_id: 1, get_client_id: 1]
   import Ecto.Changeset
   import Ecto.Query
+  import Core.ContractRequests.Storage
```
Comment (progsmile) [inline apps/core/lib/core/contract_requests/contract_requests.ex]: why import and aliasing?

Comment (PavelVesnin) [reply apps/core/lib/core/contract_requests/contract_requests.ex]: for [defdelegate](https://github.com/edenlabllc/ehealth.api/pull/3734/files#diff-aa9f80e440e676ddbf417ac576e4ef31R102)
and some functions are used by module

**Pull request merged**